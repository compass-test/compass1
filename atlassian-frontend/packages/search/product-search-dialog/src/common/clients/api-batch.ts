import debounce from 'lodash/debounce';

interface BatchMap<T, U> {
  [id: string]: {
    resolve: (response: U | PromiseLike<U>) => void;
    reject: () => void;
    arg: T;
  };
}

/**
 * A utility to batch invocations to a function.
 * All the batched requests with their arguments & returned promises are stored in the BatchMap.
 * The batched method needs a distinct `id` argument followed by the request object.
 *
 * @param {function} argumentResolver - Iterates on the BatchMap and generates the final request object which needs to be passed to the API.
 * @param {function} finalFn - The API which needs to be invoked.
 * @param {function} successfulResponseMapper - For the given id and batched successful response it extracts the response for the individual call.
 * @param {number | optional} debounceTime - The time till which the debouncing should happen.
 *
 * @typeParam {T} - It is the argument taken by the batched function.
 * @typeParam {R} - The response returned by the final function.
 * @typeParam {U} - The response type for individual method invocation.
 */
export const batchMethod = <T, R, U>(
  argumentResolver: (args: T[]) => T,
  finalFn: (arg: T) => Promise<R>,
  successfulResponseMapper: (response: R, id: string) => U,
  debounceTime: number = 100,
): ((id: string, arg: T) => Promise<U>) => {
  const batchMap: BatchMap<T, U> = {};
  const timedFn = debounce(() => {
    const inFlightRequestsMap = { ...batchMap };
    const keys = Object.keys(inFlightRequestsMap);
    const args = keys.map((key) => inFlightRequestsMap[key].arg);

    finalFn(argumentResolver(args))
      .then((response: R) => {
        keys.forEach((key) => {
          inFlightRequestsMap[key].resolve(
            successfulResponseMapper(response, key),
          );
        });
      })
      .catch(() => {
        keys.forEach((product) => {
          inFlightRequestsMap[product].reject();
        });
      })
      .finally(() => keys.forEach((key) => delete batchMap[key]));
  }, debounceTime);

  return (id: string, arg: T) => {
    const promise = new Promise<U>((resolve, reject) => {
      batchMap[id] = {
        resolve,
        reject,
        arg,
      };
    });

    timedFn();

    return promise;
  };
};
