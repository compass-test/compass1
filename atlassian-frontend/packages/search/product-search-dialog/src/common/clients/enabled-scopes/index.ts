export { getFetchProductPermissions } from './enabled-scopes';
