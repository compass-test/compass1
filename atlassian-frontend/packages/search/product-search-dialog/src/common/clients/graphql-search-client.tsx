import {
  RequestServiceOptions,
  ServiceConfig,
  utils,
} from '@atlaskit/util-service-support';
import {
  SearchContext,
  ScopedAggregatorResponse,
  ModelParam,
  Site,
} from './common-types';
import { Products } from '../product-context';
import { timed } from './timing';
import { limitToMaxSiteUsage } from './multi-site-utils';
import {
  sitesToLocations,
  SEARCH_QUERY,
  ConfluenceGraphQLFilters,
} from './graphql-tools';

export { CONFLUENCE_ITEM_ENTITIES } from './graphql-tools';

export interface SearchParams {
  query: string;
  context: SearchContext;
  entities: string[];
  modelParams: ModelParam[];
  resultLimit?: number;
  filters?: ConfluenceGraphQLFilters[];
  experience: string;
  /**
   * A list of cloud ids to search. If this is empty we will use the default `cloudId` that is used to construct this client to perform the search.
   */
  sites?: string[];
}

export interface SearchGraphQLResponse {
  // TODO : type the GraphQL response
}

export interface SearchResponseWithTiming {
  response: SearchGraphQLResponse;
  requestDurationMs: number;
}

export interface ProductsPermissionsResponse<SupportedScopes> {
  products: { [key in keyof typeof Products]: SupportedScopes[] };
}

export interface AggregatorConfig {
  cloudId: string;
  siteMasterList: Site[];
  url: string;
  useGetForScopesAPI?: boolean;
}

export type ServerProduct = 'jira' | 'confluence';

export class GraphQLClient<
  Resp extends ScopedAggregatorResponse<Scope>,
  Scope = Resp['id']
> {
  private serviceConfig: ServiceConfig;

  private cloudId: string;

  private siteMasterList: Site[];

  // Static URLs

  private static readonly GRAPHQL_API_URL = 'graphql';

  constructor({ url, cloudId, siteMasterList }: AggregatorConfig) {
    this.serviceConfig = { url: url };
    this.cloudId = cloudId;
    this.siteMasterList = siteMasterList;
  }

  public async search({
    query,
    // todo
    context,
    entities,
    // todo
    modelParams,
    // todo
    resultLimit,
    filters,
    experience,
    sites = [],
  }: SearchParams): Promise<SearchGraphQLResponse> {
    const cloudId = this.cloudId;
    const cloudIdsToFilterOn =
      sites.length > 0
        ? sites
        : limitToMaxSiteUsage(this.siteMasterList).map((site) => site.cloudId);

    const locations = sitesToLocations(cloudIdsToFilterOn, cloudId);

    const body = {
      query: SEARCH_QUERY,
      variables: {
        locations,
        entities,
        query,
        confluenceFilters: filters,
        experience,
      },
    };

    const { result: response, durationMs: requestDurationMs } = await timed(
      this.makePostRequest<SearchGraphQLResponse>(
        GraphQLClient.GRAPHQL_API_URL,
        body,
      ),
    );

    return {
      response: {
        rawData: response,
      },
      requestDurationMs,
    };
  }

  private makePostRequest = <T,>(path: string, body: object): Promise<T> => {
    const options: RequestServiceOptions = {
      path,
      requestInit: {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      },
    };
    return utils.requestService<T>(this.serviceConfig, options);
  };
}
