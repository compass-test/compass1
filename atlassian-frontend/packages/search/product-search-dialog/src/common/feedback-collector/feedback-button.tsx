import React, { FunctionComponent } from 'react';
import { FormattedMessage } from 'react-intl';
import { messages } from '../../messages';
import CustomThemeButton from '@atlaskit/button/custom-theme-button';

export interface FeedbackButtonProps {
  onClick: (e: React.MouseEvent) => any;
}

export const FeedbackButton: FunctionComponent<{
  onClick: (e: React.KeyboardEvent | React.MouseEvent) => any;
}> = ({ onClick }) => {
  return (
    <CustomThemeButton
      onClick={onClick}
      appearance="subtle"
      theme={(theme: any, props: any) => {
        const currentStyles = theme({ ...props });
        return {
          ...currentStyles,
          buttonStyles: {
            ...currentStyles.buttonStyles,
            margin: '0 auto',
            display: 'block',
          },
        };
      }}
    >
      <FormattedMessage {...messages.common_give_feedback} />
    </CustomThemeButton>
  );
};
