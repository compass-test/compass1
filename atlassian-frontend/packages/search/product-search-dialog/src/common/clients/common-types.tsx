import { FilterType as ConfluenceFilterType } from '../../confluence/clients';
import { FilterType as JiraFilterType } from '../../jira/clients';
import { Products } from '../product-context';

export interface ModelParam {
  '@type': string;
  [value: string]: string | number;
}

export interface SearchContext {
  sessionId: string;
  referrerId: string | null;
}

export interface ABTest {
  abTestId: string;
  controlId: string;
  experimentId: string;
}

export interface ScopedAggregatorResponse<T> {
  id: T;
  error?: string;
  size?: number;
}

export interface ExperimentResponse<T> {
  scopes: {
    id: T;
    abTest: ABTest;
  }[];
}

export type CommonFilterType = 'site';

export type ProductFilterType =
  | ConfluenceFilterType
  | JiraFilterType
  | CommonFilterType;

/**
 * All filters must implement this interface.
 */
export interface FilterInterface {
  '@type': string;
}

/****************************************************************/
/**                   Collaboration Graph                      **/
/****************************************************************/

export const SpaceContainerType = 'confluenceSpace';
export const ProjectContainerType = 'jiraProject';

export type ContainerType =
  | typeof SpaceContainerType
  | typeof ProjectContainerType;

export interface CollaborationGraphContainerAPIResponse {
  entityType: string;
  containerType: ContainerType;
  id: string;
  containerDetails: {
    id: string;
    key: string;
    name: string;
    url: string;
    iconUrl: string;
    score: number;
  };
}

export interface CollaborationGraphContainer
  extends CollaborationGraphContainerAPIResponse {
  siteId: string;
}

export interface CollaborationGraphUserAPIResponse {
  entityType: string;
  id: string;
  userProfile: {
    account_id: string;
    email: string;
    name: string;
    picture: string;
    account_status: string;
    account_type: string;
    locale: string;
    score: number;
  };
}

export interface CollaborationGraphUser
  extends CollaborationGraphUserAPIResponse {
  siteId: string;
}

export interface CollaborationGraphResponse<T> {
  collaborationGraphEntities: T[];
  timings: number;
}

export interface SharedClient {
  getAbTestData: () => Promise<ABTest | null>;
  getProductPermissions: (products: Products[]) => Promise<Products[]>;
}

/****************************************************************/
/**                   Multi Site                               **/
/****************************************************************/

/**
 * Defines the properties to be passed to serve results from multiple tenants.
 *
 * @siteUrl {string} can be thought of as the hostUrl without a trailing '/'. Not to be confused with 'baseUrl' which has the value '/wiki' for confluence.
 */
export interface Site {
  product: Products;
  avatarUrl: string;
  cloudId: string;
  collabGraphSessionId?: string;
  siteName: string;
  siteUrl: string;
}
