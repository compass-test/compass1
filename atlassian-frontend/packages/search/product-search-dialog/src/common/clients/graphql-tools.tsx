import { Filter } from '../../confluence/clients/confluence-search-client';

export const CONFLUENCE_ITEM_ENTITIES = [
  'ati:cloud:confluence:page',
  'ati:cloud:confluence:attachment',
  'ati:cloud:confluence:blogpost',
];

const SEARCH_QUERY = `
query($query: String, $entities:[String!]!, $locations: [String!]!, $confluenceFilters: SearchConfluenceFilter, $experience: String!){
  search{
    search(
      query: $query
      filters: {entities: $entities, locations: $locations, confluenceFilters: $confluenceFilters}
      analytics: {searchSessionId: "graphql"}
      experience: $experience
    ) {
      edges {
        node {
          id
          title
          url
          description
          type
          lastModifiedDate
          __typename
          ... on SearchConfluencePageBlogAttachment {
            type
            space {
              key
              iconUrl
            }
          }
        }
      }
    }
  }
}`;

const sitesToLocations = (sites: string[], cloudId: string): String[] => {
  const cloudIds = sites.length > 0 ? sites : [cloudId];
  const locations = cloudIds.map(
    (cloudId) => `ari:cloud:confluence::site/${cloudId}`,
  );
  return locations;
};

export interface ConfluenceGraphQLFilters {
  spaceFilter?: string[];
  contributorsFilter?: string[];
}

const confluenceFilterToGraphQLFilters = (
  filters: Filter[],
): ConfluenceGraphQLFilters => {
  return {
    spaceFilter: filters
      .filter((filter) => filter.type === 'space')
      .map((filter) => filter.id),
    contributorsFilter: filters
      .filter((filter) => filter.type === 'contributor')
      .map((filter) => filter.id),
  };
};

export { sitesToLocations, confluenceFilterToGraphQLFilters, SEARCH_QUERY };
