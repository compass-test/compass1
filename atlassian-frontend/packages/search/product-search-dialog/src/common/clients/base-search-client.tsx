import {
  RequestServiceOptions,
  ServiceConfig,
  utils,
} from '@atlaskit/util-service-support';
import {
  SearchContext,
  ABTest,
  ScopedAggregatorResponse,
  ModelParam,
  Site,
} from './common-types';
import { Products } from '../product-context';
import { timed } from './timing';
import { limitToMaxSiteUsage } from './multi-site-utils';
import { batchMethod } from './api-batch';
import { getFetchProductPermissions } from './enabled-scopes';

export interface SearchParams<SearchScopes, SupportedFilters> {
  query: string;
  context: SearchContext;
  scopes: SearchScopes;
  modelParams: ModelParam[];
  resultLimit?: number;
  filters?: SupportedFilters[];
  experience: string;
  /**
   * A list of cloud ids to search. If this is empty we will use the default `cloudId` that is used to construct this client to perform the search.
   */
  sites?: string[];
}

// Type used to help unpack the types that make up an array
type Unpack<A> = A extends Array<infer E> ? E : A;
// Type used to lookup the response based on the scope
type ResponseFromScope<S, T> = S extends ScopedAggregatorResponse<infer E>
  ? T extends E
    ? S
    : never
  : never;

interface MultiscopeSearchResponse<T> {
  scopes: T[];
}

export interface SearchResponse<
  Scope,
  Response extends ScopedAggregatorResponse<Scope>,
  SearchScope
> {
  rawData: MultiscopeSearchResponse<
    ResponseFromScope<Response, Unpack<SearchScope>>
  >;
  retrieveScope<SingleScope extends Unpack<SearchScope>>(
    scope: SingleScope,
  ): ResponseFromScope<Response, SingleScope> | null;
}

export interface SearchResponseWithTiming<
  S,
  R extends ScopedAggregatorResponse<S>,
  SR
> {
  response: SearchResponse<S, R, SR>;
  requestDurationMs: number;
}

interface ScopedAbTestResponse<SupportedScopes> {
  id: SupportedScopes;
  error?: string;
  abTest: ABTest;
}

interface MultiscopeAbTestResponse<SupportedScopes> {
  scopes: ScopedAbTestResponse<SupportedScopes>[];
}

export interface ProductsPermissionsResponse<SupportedScopes> {
  products: { [key in keyof typeof Products]: SupportedScopes[] };
}

export interface AggregatorConfig {
  cloudId: string;
  siteMasterList: Site[];
  url: string;
  useGetForScopesAPI?: boolean;
}

export type ServerProduct = 'jira' | 'confluence';

interface ProductPermissions {
  productIds: string[];
  experience: string;
}

export class AggregatorClient<
  Resp extends ScopedAggregatorResponse<Scope>,
  SupportedFilters,
  Scope = Resp['id']
> {
  private serviceConfig: ServiceConfig;

  private cloudId: string;

  private siteMasterList: Site[];

  // Static URLs
  private static readonly EXPERIMENT_URL_PATH = 'experiment/v1';

  private static readonly QUICKSEARCH_API_URL = 'quicksearch/v1';

  constructor({ url, cloudId, siteMasterList }: AggregatorConfig) {
    this.serviceConfig = { url: url };
    this.cloudId = cloudId;
    this.siteMasterList = siteMasterList;
  }

  public async getAbTestData(
    scope: Scope,
    experience: string,
  ): Promise<ABTest | null> {
    const body = {
      cloudId: this.cloudId,
      scopes: [scope],
      experience,
    };

    const response = await this.makePostRequest<
      MultiscopeAbTestResponse<Scope>
    >(AggregatorClient.EXPERIMENT_URL_PATH, body);

    const scopeWithAbTest:
      | ScopedAbTestResponse<Scope>
      | undefined = response.scopes.find((s) => s.id === scope);

    const abTestPromise = scopeWithAbTest
      ? Promise.resolve(scopeWithAbTest.abTest)
      : Promise.resolve(null);

    return abTestPromise;
  }

  /**
   * Get a list of permitted scopes for the supplied product.
   *
   * Consumers should favour {@link batchedGetExtensibleProductPermission}, to batch
   * multiple scopes calls into one. {@link getProductScopes} may be used when a caller
   * explicitly wants a separate REST call to fetch scopes. One example may be if a product
   * has different scopes caching rules to other products.
   *
   * @param productId the product id to lookup scopes for
   * @param experience the experience this UI represents
   * @returns a promise of a list of scopes for the supplied product
   */
  public async getProductScopes(
    productId: string,
    experience: string,
  ): Promise<string[]> {
    const productScopeMap = await this.getGenericProductPermissions({
      productIds: [productId],
      experience,
    });

    return productScopeMap.get(productId) || [];
  }

  /**
   * Returns a list of {@link Products} which the user should see in the dialog
   *
   * @deprecated consumers should use {@link batchedGetExtensibleProductPermission} to fetch product scopes, or
   * {@link getProductScopes} to force a separate call to fetch a single product's scopes.
   *
   */
  public async getProductPermissions(
    products: Products[],
    experience: string,
  ): Promise<Products[]> {
    const lowercaseProductIds = products.map((p) => p.toLowerCase());

    const productScopeMap = await this.getGenericProductPermissions({
      productIds: lowercaseProductIds,
      experience,
    });

    return lowercaseProductIds
      .filter((product) => productScopeMap.get(product)?.length)
      .map((product) => product.toUpperCase() as Products);
  }

  private getGenericProductPermissions = async (
    args: ProductPermissions,
  ): Promise<Map<string, string[]>> => {
    return getFetchProductPermissions(this.makeGetRequest)(
      args.productIds,
      this.cloudId,
      args.experience,
    );
  };

  public batchedGetExtensibleProductPermission = batchMethod<
    ProductPermissions,
    Map<string, string[]>,
    string[]
  >(
    (args) => {
      const emptyProductPermissions = {
        productIds: [],
        experience: '',
      };

      if (args.length === 0) {
        return emptyProductPermissions;
      }

      return args.reduce<ProductPermissions>((acc, arg) => {
        return {
          productIds: [...acc.productIds, ...arg.productIds],
          experience: arg.experience,
        };
      }, emptyProductPermissions);
    },
    this.getGenericProductPermissions,
    (response, id) => response.get(id) || [],
  );

  public async search<SearchScopes extends Scope[]>({
    query,
    context,
    scopes,
    modelParams,
    resultLimit,
    filters = [],
    experience,
    sites = [],
  }: SearchParams<SearchScopes, SupportedFilters>): Promise<
    SearchResponseWithTiming<Scope, Resp, SearchScopes>
  > {
    const cloudId = this.cloudId;
    const cloudIdsToFilterOn =
      sites.length > 0
        ? sites
        : limitToMaxSiteUsage(this.siteMasterList).map((site) => site.cloudId);
    const body = {
      query: query,
      cloudId,
      limit: resultLimit,
      scopes,
      filters: filters,
      searchSession: context,
      ...(modelParams.length > 0 ? { modelParams } : {}),
      experience,
      cloudIds: cloudIdsToFilterOn,
    };

    const { result: response, durationMs: requestDurationMs } = await timed(
      this.makePostRequest<
        MultiscopeSearchResponse<ResponseFromScope<Resp, Unpack<SearchScopes>>>
      >(AggregatorClient.QUICKSEARCH_API_URL, body),
    );

    return {
      response: {
        rawData: response,
        retrieveScope: (scope) =>
          response.scopes.find((s) => s.id === scope) || null,
      },
      requestDurationMs,
    };
  }

  private makePostRequest = <T,>(path: string, body: object): Promise<T> => {
    const options: RequestServiceOptions = {
      path,
      requestInit: {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      },
    };
    return utils.requestService<T>(this.serviceConfig, options);
  };

  private makeGetRequest = <T,>(path: string): Promise<T> => {
    const options: RequestServiceOptions = {
      path,
      requestInit: {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      },
    };
    return utils.requestService<T>(this.serviceConfig, options);
  };
}
