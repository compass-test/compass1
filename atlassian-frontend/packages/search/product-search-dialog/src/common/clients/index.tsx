import * as CommonTypes from './common-types';

export { AggregatorClient } from './base-search-client';
export { CollaborationGraphClient } from './collaboration-graph-client';
export { responseErrorToError } from './error-handler';
export { extractSiteUrl } from './multi-site-utils';
export { timed } from './timing';

export type ModelParam = CommonTypes.ModelParam;
export type SearchContext = CommonTypes.SearchContext;
export type AggregatorConfig = import('./base-search-client').AggregatorConfig;
export type ProductsPermissionsResponse<
  T
> = import('./base-search-client').ProductsPermissionsResponse<T>;
export type CollaborationGraphConfig = import('./collaboration-graph-client').CollaborationGraphConfig;
export type ABTest = CommonTypes.ABTest;
export type ContainerType = CommonTypes.ContainerType;
export type ScopedAggregatorResponse<T> = CommonTypes.ScopedAggregatorResponse<
  T
>;
export type CollaborationGraphResponse<
  T
> = CommonTypes.CollaborationGraphResponse<T>;
export type CollaborationGraphContainer = CommonTypes.CollaborationGraphContainer;
export type SharedClient = CommonTypes.SharedClient;

// Multi site
export type Site = CommonTypes.Site;
