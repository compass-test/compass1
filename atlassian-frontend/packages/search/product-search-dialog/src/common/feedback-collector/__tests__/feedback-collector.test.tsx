import React from 'react';
import { mount } from 'enzyme';
import { act } from '@testing-library/react';
import AkFeedbackCollector, {
  FeedbackFlag,
} from '@atlaskit/feedback-collector';
import FeedbackCollector from '../feedback-collector';
import { ABTest } from '../../clients';
import { FeedbackButton } from '../feedback-button';

jest.mock('@atlaskit/feedback-collector', () => ({
  __esModule: true,
  default: () => <div data-test-id="AkFeedbackCollector" />,
  FeedbackFlag: () => <div data-test-id="FeedbackFlag" />,
}));

jest.mock('@atlaskit/button', () => () => <div />);

jest.mock('../feedback-button', () => ({
  FeedbackButton: () => <div data-test-id="FeedbackButton" />,
}));

jest.mock('../../ab-test-provider', () => ({
  useABTest: () =>
    ({
      abTestId: 'some_ab_test',
      experimentId: 'some_experiment',
      controlId: 'some_control',
    } as ABTest),
}));

jest.mock('../../user-context', () => ({
  useUserContext: () => ({
    user: {
      name: 'some_name',
      email: 'some_email',
    },
  }),
}));

jest.mock('../../product-context', () => {
  const productContext = jest.requireActual('../../product-context');
  return Object.assign({}, productContext, {
    usePrimaryProduct: () => productContext.Products.confluence,
    useActiveProduct: () => productContext.Products.jira,
  });
});

describe('<FeedbackCollector />', () => {
  it('renders the feedback button by default', () => {
    const wrapper = mount(<FeedbackCollector />);

    expect(wrapper.find(FeedbackFlag).exists()).toBe(false);
    expect(wrapper.find(AkFeedbackCollector).exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the feedback collector dialog if button is clicked', () => {
    const wrapper = mount(<FeedbackCollector />);

    act(() => {
      wrapper.find(FeedbackButton).prop('onClick')!({
        preventDefault: jest.fn(),
      } as any);
    });

    wrapper.update();

    expect(wrapper.find(FeedbackFlag).exists()).toBe(false);
    expect(wrapper.find(AkFeedbackCollector).exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the feedback flag and closes dialog if form is submitted', () => {
    const wrapper = mount(<FeedbackCollector />);

    act(() => {
      // Open the form
      wrapper.find(FeedbackButton).prop('onClick')!({
        preventDefault: jest.fn(),
      } as any);
    });

    wrapper.update();

    act(() => {
      // The feedback collector fires both onSubmit and onClose separately when submit is clicked
      wrapper.find(AkFeedbackCollector).prop('onSubmit')!({
        type: 'comment',
        description: 'Nice!',
        canBeContacted: true,
        enrollInResearchGroup: true,
      });
      wrapper.find(AkFeedbackCollector).prop('onClose')!();
    });

    wrapper.update();

    expect(wrapper.find(FeedbackFlag).exists()).toBe(true);
    expect(wrapper.find(AkFeedbackCollector).exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render the feedback flag if form is cancelled', () => {
    const wrapper = mount(<FeedbackCollector />);

    act(() => {
      // Open the form
      wrapper.find(FeedbackButton).prop('onClick')!({
        preventDefault: jest.fn(),
      } as any);
    });

    wrapper.update();

    act(() => {
      // The feedback collector fires onClose only if cancelled
      wrapper.find(AkFeedbackCollector).prop('onClose')!();
    });

    wrapper.update();

    expect(wrapper.find(FeedbackFlag).exists()).toBe(false);
    expect(wrapper.find(AkFeedbackCollector).exists()).toBe(false);
  });
});
