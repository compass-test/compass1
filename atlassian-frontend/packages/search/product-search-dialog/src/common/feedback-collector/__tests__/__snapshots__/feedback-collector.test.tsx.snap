// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedbackCollector /> renders the feedback button by default 1`] = `
<FeedbackCollector>
  <FeedbackButton
    onClick={[Function]}
  >
    <div
      data-test-id="FeedbackButton"
    />
  </FeedbackButton>
</FeedbackCollector>
`;

exports[`<FeedbackCollector /> renders the feedback collector dialog if button is clicked 1`] = `
<FeedbackCollector>
  <FeedbackButton
    onClick={[Function]}
  >
    <div
      data-test-id="FeedbackButton"
    />
  </FeedbackButton>
  <_default
    additionalFields={
      Array [
        Object {
          "id": "customfield_10047",
          "value": "experimentId: some_experiment, abTestId: some_ab_test, frontendVersion: @atlassian/product-search-dialog@__BUILD_VERSION__, product: jira, primaryProduct: confluence",
        },
      ]
    }
    email="some_email"
    embeddableKey="85dc6027-c074-4800-ba54-4ecb844b29f8"
    name="some_name"
    onClose={[Function]}
    onSubmit={[Function]}
    requestTypeId="182"
  >
    <div
      data-test-id="AkFeedbackCollector"
    />
  </_default>
</FeedbackCollector>
`;

exports[`<FeedbackCollector /> renders the feedback flag and closes dialog if form is submitted 1`] = `
.emotion-2 {
  bottom: 48px;
  left: 80px;
  position: fixed;
  z-index: 600;
}

@media (max-width:560px) {
  .emotion-2 {
    bottom: 0;
    left: 0;
  }
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-1 {
  bottom: 0;
  position: absolute;
  width: 400px;
  -webkit-transition: -webkit-transform 400ms ease-in-out;
  -webkit-transition: transform 400ms ease-in-out;
  transition: transform 400ms ease-in-out;
}

@media (max-width:560px) {
  .emotion-1 {
    width: 100vw;
  }
}

.emotion-1:first-of-type {
  -webkit-transition: none;
  transition: none;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

.emotion-1:nth-of-type(n + 2) {
  -webkit-animation-duration: 0ms;
  animation-duration: 0ms;
  -webkit-transform: translateX(0) translateY(100%) translateY(16px);
  -ms-transform: translateX(0) translateY(100%) translateY(16px);
  transform: translateX(0) translateY(100%) translateY(16px);
}

.emotion-1:nth-of-type(1) {
  z-index: 5;
}

.emotion-1:nth-of-type(2) {
  z-index: 4;
}

.emotion-1:nth-of-type(n + 4) {
  visibility: hidden;
}

.emotion-1.emotion-1 + * {
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
  -webkit-transition-duration: 200ms;
  transition-duration: 200ms;
}

<FeedbackCollector>
  <FeedbackButton
    onClick={[Function]}
  >
    <div
      data-test-id="FeedbackButton"
    />
  </FeedbackButton>
  <FlagGroup
    onDismissed={[Function]}
  >
    <Portal
      zIndex={600}
    >
      <InternalPortal
        zIndex={600}
      >
        <Portal
          containerInfo={
            .emotion-2 {
  bottom: 48px;
  left: 80px;
  position: fixed;
  z-index: 600;
}

@media (max-width:560px) {
  .emotion-2 {
    bottom: 0;
    left: 0;
  }
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-1 {
  bottom: 0;
  position: absolute;
  width: 400px;
  -webkit-transition: -webkit-transform 400ms ease-in-out;
  -webkit-transition: transform 400ms ease-in-out;
  transition: transform 400ms ease-in-out;
}

@media (max-width:560px) {
  .emotion-1 {
    width: 100vw;
  }
}

.emotion-1:first-of-type {
  -webkit-transition: none;
  transition: none;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

.emotion-1:nth-of-type(n + 2) {
  -webkit-animation-duration: 0ms;
  animation-duration: 0ms;
  -webkit-transform: translateX(0) translateY(100%) translateY(16px);
  -ms-transform: translateX(0) translateY(100%) translateY(16px);
  transform: translateX(0) translateY(100%) translateY(16px);
}

.emotion-1:nth-of-type(1) {
  z-index: 5;
}

.emotion-1:nth-of-type(2) {
  z-index: 4;
}

.emotion-1:nth-of-type(n + 4) {
  visibility: hidden;
}

.emotion-1.emotion-1 + * {
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
  -webkit-transition-duration: 200ms;
  transition-duration: 200ms;
}

<div
              class="atlaskit-portal"
              style="z-index: 600;"
            >
              <div
                class="emotion-2"
              >
                <h2
                  class="emotion-0"
                >
                  Flag notifications
                </h2>
                <div
                  class=" emotion-1"
                >
                  <div
                    data-test-id="FeedbackFlag"
                  />
                </div>
              </div>
            </div>
          }
        >
          <div
            className="emotion-2"
          >
            <h2
              className="emotion-0"
            >
              Flag notifications
            </h2>
            <ExitingPersistence
              appear={false}
            >
              <SlideIn
                animationTimingFunction={[Function]}
                duration={400}
                enterFrom="left"
                fade="inout"
                key=".$.0"
              >
                <EnteringMotion
                  animationTimingFunction={[Function]}
                  duration={400}
                  enteringAnimation={
                    Object {
                      "0%": Object {
                        "opacity": 0,
                        "transform": "translate3d(-100%, 0, 0)",
                      },
                      "100%": Object {
                        "opacity": 1,
                        "transform": "none",
                      },
                    }
                  }
                  exitingAnimation={
                    Object {
                      "0%": Object {
                        "opacity": 1,
                        "transform": "none",
                      },
                      "100%": Object {
                        "opacity": 0,
                        "transform": "translate3d(-100%, 0, 0)",
                      },
                    }
                  }
                >
                  <ForwardRef(render)>
                    <div
                      className=" emotion-1"
                    >
                      <FeedbackFlag>
                        <div
                          data-test-id="FeedbackFlag"
                        />
                      </FeedbackFlag>
                    </div>
                  </ForwardRef(render)>
                </EnteringMotion>
              </SlideIn>
            </ExitingPersistence>
          </div>
        </Portal>
      </InternalPortal>
    </Portal>
  </FlagGroup>
</FeedbackCollector>
`;
